core/container/avl/avl.odin

```diff
@@ -0,0 +1,598 @@
+/*
+package avl implements an AVL tree.
+
+The implementation is non-intrusive, and non-recursive.
+*/
+package container_avl
+
+import "base:runtime"
+_ :: runtime
+
+// Originally based on the CC0 implementation by Eric Biggers
+// See: https://github.com/ebiggers/avl_tree/
+
+// Direction specifies the traversal direction for a tree iterator.
+Direction :: enum i8 {
+	// Backward is the in-order backwards direction.
+	Backward = -1,
+	// Forward is the in-order forwards direction.
+	Forward  = 1,
+}
+
+// Tree is an AVL tree.
+Tree :: struct($Value: typeid) {
+	on_remove:           proc(value: Value, user_data: rawptr),
+	on_remove_user_data: rawptr,
+
+	_root:               ^Node(Value),
+	_node_allocator:     runtime.Allocator,
+	_cmp_fn:             proc(a, b: Value) -> int,
+	_size:               int,
+}
+
+// Node is an AVL tree node.
+//
+// WARNING: It is unsafe to mutate value if the node is part of a tree
+// if doing so will alter the Node's sort position relative to other
+// elements in the tree.
+Node :: struct($Value: typeid) {
+	value:    Value,
+
+	_parent:  ^Node(Value),
+	_left:    ^Node(Value),
+	_right:   ^Node(Value),
+	_balance: i8,
+}
+
+// Iterator is a tree iterator.
+//
+// WARNING: It is unsafe to modify the tree while iterating, except via
+// the iterator_remove method.
+Iterator :: struct($Value: typeid) {
+	_tree:        ^Tree(Value),
+	_cur:         ^Node(Value),
+	_next:        ^Node(Value),
+	_direction:   Direction,
+	_initialized: bool,
+}
+
+// init initializes a tree.  The comparision function MUST return
+// < 0, 0, or > 0 if a < b, a == b, or a > b respectively.
+init :: proc(
```

Comment:
To make it easier to use, a different initialiser for comparable types would be nice (which sets the cmp_proc to a `a - b`.

Suggestion:
```odin
init :: proc {
  init_comparable,
  init_cmp,
}

init_comparable :: proc(t: ^Tree($Value), node_allocator := context.allocator)
	where intrinsics.type_is_comparable(Value) {
	init_cmp(t, proc(a, b: Value) -> int { return a - b }, node_allocator)
}
```

and rename the current `init` to `init_cmp`.